### ISSUE TITLE
Support X11 forwarding

### ISSUE THREAD

**Issue by fpoirotte** [loft-sh/devpod#654]
**Is your feature request related to a problem?**  
Hi, I would like to run a graphical application inside a devpod workspace, using X11 forwarding.

I found no information on how to enable the X11 forwarding protocol in the documentations.

Is this already supported? If not, is it a planned feature?
What would be the best way to achieve X11 forwarding?

**Which solution do you suggest?**  
If the feature is already available, add a quickstart how-to on the subject.
If not, add support for X11 forwarding.

**Which alternative solutions exist?**  
None that I am aware of.

**Additional context**  
n/a

**Comment by pascalbreuninger** [comment]
Hi @fpoirotte, just saw this issue now. Supporting X11 is on our roadmap, albeit not with the highest priority. 
If you need it quickly, feel free to open a PR and kickstart the feature

**Comment by decipher2k** [comment]
X11 forwarding would be a killer feature. DevPod isn't really cool without it.

**Comment by decipher2k** [comment]
> **Is your feature request related to a problem?** Hi, I would like to run a graphical application inside a devpod workspace, using X11 forwarding.
> 
> I found no information on how to enable the X11 forwarding protocol in the documentations.
> 
> Is this already supported? If not, is it a planned feature? What would be the best way to achieve X11 forwarding?
> 
> **Which solution do you suggest?** If the feature is already available, add a quickstart how-to on the subject. If not, add support for X11 forwarding.
> 
> **Which alternative solutions exist?** None that I am aware of.
> 
> **Additional context** n/a

You could setup your own workspace and use Xming:
http://www.straightrunning.com/XmingNotes/
It's an X11-Server for Windows.

**Comment by behrica** [comment]
https://github.com/novnc/noVNC

is an other option to run linux graphical apps inside Docker and expose a "web browser interface" to them.
This should work as-is, as it iust needs a TCP port exposed.

**Comment by vquemener** [comment]
> Hi @fpoirotte, just saw this issue now. Supporting X11 is on our roadmap, albeit not with the highest priority. If you need it quickly, feel free to open a PR and kickstart the feature

Hi @pascalbreuninger, thanks for your awesome work on DevPod!

Is this roadmap public?

Would you say there's a good chance that X11 forwarding support will be available in 2024?

**Comment by pascalbreuninger** [comment]
@vquemener thanks :) 

Nope, it's in our internal tracker but we were discussing opening the DevPod Roadmap up to the open source community. No tangible outcome yet though

As for X11 support, it's certainly possible in 2024 but your contribution could even bring it in 2023 ;)

**Comment by fpoirotte** [comment]
I've created draft pull requests to address this issue.
I'm a novice when it comes to Go, so apologies if the code quality is subpar.
I haven't had time to write tests yet (I'm not entirely sure how to test this anyway as e2e tests would require a working graphical server for this to work).

**Comment by pascalbreuninger** [comment]
@fpoirotte thanks a lot, we'll take a look!

**Comment by haakobja** [comment]
> **Is your feature request related to a problem?** Hi, I would like to run a graphical application inside a devpod workspace, using X11 forwarding.
> 
> I found no information on how to enable the X11 forwarding protocol in the documentations.
> 
> Is this already supported? If not, is it a planned feature? What would be the best way to achieve X11 forwarding?
> 
> **Which solution do you suggest?** If the feature is already available, add a quickstart how-to on the subject. If not, add support for X11 forwarding.
> 
> **Which alternative solutions exist?** None that I am aware of.
> 
> **Additional context** n/a

I know this is an old topic, but here's my 5 cents.

X11 Forwarding is possible by adding the `DISPLAY` environment variable to devcontainer.json. However it only works without extra software if you're using Linux. StackOverflow provides more information: https://stackoverflow.com/questions/44429394/x11-forwarding-of-a-gui-app-running-in-docker.

If you're using Windows, I don't know how to get it to work, but I successfully got it to work on Mac using XQuartz and adding this to my `devcontainer.json`:

```json
"containerEnv": {
  "DISPLAY": "host.containers.internal:0"
}
```

Please note that XQuartz must be running and you must execute the following command `xhost + 127.0.0.1` before you start your container

**Comment by fpoirotte** [comment]
@haakobja : yes, it's possible to set the `DISPLAY` variable environment, however it exposes the application to additional risk: the X11 traffic will not pass through devpod's SSH tunnel (and is thus sent in cleartext).

It also adds additional constraints (e.g. the X11 server must listen on the network rather than on a UNIX socket, or you must use the host's network inside the container, or mount the X11 server's UNIX socket inside the container and authorize the container explicitly with tools like `xauth`).
All of these solutions are cumbersome and may pose additional security problems.

The idea behind this ticket is to bake support for the SSH X11 forwarding protocol into devpod's SSH server directly.
That way, users can simply enable X11 forwarding in their SSH client when connecting to the workspace (`ssh -X workspace.devpod`) and graphical applications will "just work" provided there is a running X11 server on the client.
